server/admin/http/models.go

```diff
@@ -53,3 +56,20 @@ type (
 		Paths []string `json:"paths"`
 	}
 )
+
+type (
+	// AnnounceAdvReq represents a request for publishing a raw advertisement.
+	AnnounceAdvReq struct {
+		Indices         []provider.IteratorStep
```

Comment:
If this is meant to be passed over an HTTP interface, I think `Indices` would be better as type `[]multihash.Multihash`, since there is no need for the offsets. The slice of `multihash.Multihash` should be correctly ordered by the client.

Also, since you do not need the offsets, there is no need to reuse the existing `MultihashIterator` implementation `indexMhIterator`.  Better to create your own `MultihashIterator` implementation.  Maybe it could look something like:
```go
type indexMhIterator struct {                                                                                                                
    mhs []multihash.Multihash                                                                                                                
    pos int                                                                                                                                  
}                                                                                                                                            
func (i *indexMhIterator) Next() (multihash.Multihash, error) {                                                                              
    // We have reached the end of stream.                                                                                                    
    if i.pos >= len(i.mhs) {                                                                                                                 
        return nil, io.EOF                                                                                                                   
    }                                                                                                                                        
    mh := i.mhs[i.pos]                                                                                                                       
    i.pos++                                                                                                                                  
    return mh, nil                                                                                                                           
}                                                                                                                                            
```
This would be constructed in ` publishAdvHandler()` :
```go
mhIter = indexMhIterator{
    mhs: req.Indices,
}
```

Reply:
make sense. i will fix this.

Reply:
> Thanks for the contribution @zl03jsj !
> 
> I wonder if there's a way to do this that would help us have a bit less additional code:
> 
> * you introduce a new struct type, `AnnounceAdvReq` on the admin server for data
> * This looks a lot like what lotus uses in its library interface to [`Publish`](https://github.com/filecoin-project/index-provider/blob/main/engine/engine.go#L191) - do you think we could re-use that type and method as what the handler calls?

i think it's worth a try. and after some researching on this, 
if we re-use  the `Engine.Publish` method , it is have to create `schema.Advertisement` out side the `engine`, 
```go
	Advertisement struct {
		PreviousID *ipld.Link
		Provider   string
		Addresses  []string
		Signature  []byte
		Entries    ipld.Link
		ContextID  []byte
		Metadata   []byte
		IsRm       bool
	}
```
to fill the its members, such as:`PreviousID`, `Entries`, `Signature`, 
the `Engine` have to expose some **protected** members outside(e.g:`key`, `entriesChunker`..), 
https://github.com/filecoin-project/index-provider/blob/cf3e8e80afdc62bf0ce94cfcb375f73e751b50cb/engine/engine_api_test.go#L19-L26

they are now only for test purpose. do your think is that ok?

and on the other side,  in the old approach, 
we just needs minor change on `PublishAdvForIndex` and make it public is ok.

i prefer expose `PublishAdvForIndex`.  (:-

Reply:
@zl03jsj the rationale behind having the entires chunks in HTTP terms is simply pagination. The HTTP endpoint wherever it ends up at will need some form of pagination for the multihash entries since they can certainly be very long. 

I recommend thinking about what the pagination substitute would be for the HTTP endpoint.